std/stdio.d

```diff
@@ -1522,6 +1524,139 @@ for every line.
+
+The file is created with R/W permissions and opened with $(D "w+b"). On POSIX
+systems, the permissions are restricted to the current user, though the
+effective permissions are modified by the process' umask in the usual way.
+
+Params:
+    prefix = Prefix for the generated file name.
+    suffix = Suffix for the generated file name (which also provides a way to
+             give the file an extension).
+    dir    = Directory of the temporary file. Defaults to the result of
+             $(XREF file, tempDir).
+  */
+    static File tempFile(string prefix = null, string suffix = null)
+    {
+        import std.file : tempDir;
+        return File.tempFile(prefix, suffix, std.file.tempDir());
+    }
+
+    /// Ditto
+    static File tempFile(string prefix, string suffix, string dir)
+    {
+        import std.exception : ErrnoException;
+        import std.file : isSymlink;
+        import std.path : absolutePath, baseName, buildPath, dirName;
+
+        static string genTempName(string prefix, string suffix, string dir)
+        {
+            import std.ascii : digits, letters;
+            import std.random : rndGen;
+            import std.range : chain;
+            auto name = new char[](prefix.length + 15 + suffix.length);
+            name[0 .. prefix.length] = prefix;
+            name[$ - suffix.length .. $] = suffix;
+
+            rndGen.popFront();
+            auto chars = chain(cast(immutable(ubyte[]))letters, cast(immutable(ubyte[]))digits);
+            foreach(ref c; name[prefix.length .. $ - suffix.length])
+            {
+                c = chars[rndGen.front % chars.length];
+                rndGen.popFront();
+            }
+
+            return buildPath(dir, name);
+        }
+
+        // Once should be enough given the randomness, but we'll give it three
+        // times to be safe.
+        foreach(i; 0 .. 3)
+        {
+            auto filename = genTempName(prefix, suffix, dir);
+
+            // O_EXCL does not support symbolic links.
+            if(dirName(filename).isSymlink)
+                filename = buildPath(absolutePath(dirName(filename), baseName(filename)));
+
+            import std.internal.cstring : tempCString;
+            version(Posix)
+            {
+                import core.sys.posix.sys.stat, core.sys.posix.fcntl;
+                enum flags = S_IRUSR | S_IWUSR;
+                auto fd = open(tempCString(filename), O_RDWR | O_CREAT | O_EXCL, flags);
+            }
+            else version(Windows)
+            {
+                enum flags = _S_IREAD | _S_IWRITE;
+                auto fd = _wsopen(tempCString!wchar(filename), O_RDWR | O_CREAT | O_EXCL, _SH_DENYNO, flags);
+            }
+            else static assert("Unsupported OS");
+
+            if(fd == -1)
+                continue;
+
+            File retval;
+            retval.fdopen(fd, "w+b", filename);
+            return retval;
+        }
+
+        throw new ErrnoException("Failed to create a temporary file.");
```

Comment:
I thought Posix used `errno` and Windows used `GetLastError` or similar. I see that's not the case. Anyway, here are a couple of reasons why using `ErrnoException` is bad:

* It leaks an implementation detail, i.e. the C error handling mechanism called from an internal function. A D user should not need to know or care about C error codes unless explicitly using C API's

* Since `ErrnoException` is used for a bunch of completely different reasons it's hard to know from where the exception is thrown. You also need to catch it to actually know if you can handle it or not

* It makes it too easy to not really care and think about error handling. Just throw `ErrnoException` instead of creating a new exception type with proper data

* It can not contain any additional information except the original error code. In this case including the filename that was generated and the reason why it failed, if possible, would be useful

* Since the filename is not included in the exception you're making it harder for the user to handle errors when something goes wrong. It's possible to extract the filename from the returned file object but that's not always the case, example:

```d
File file1;
File file2;

try {
    file1 = File.tempFile();
    file2 = File.tempFile();
} catch (ErrnoException e) {
    // here we don't know which of the to above calls failed
    // making it impossible to get the filename (as far as I can see)
}
```

The user of an application could have chosen a directory he/she doesn't have write permissions to, to create the temp file in.

An API should be primarily written to be easy for the user, not the implementer. I would say that the most uses cases of `ErrnoException` and especially together with `errnoEnforce` in std.stdio is laziness. The implementer has not bother creating the proper exception type to indicate the reason for a failed operation and include the necessary data to handle the exception.